transformations/yes_no_question/transformation.py

```diff
@@ -0,0 +1,203 @@
+from initialize import spacy_nlp
+import pyinflect
+from nltk.tokenize.treebank import TreebankWordDetokenizer
+from spacy.symbols import aux, cc, nsubj, AUX, NOUN, PRON, PROPN, VERB
+from spacy.tokens import Token, Span
+from spacy.tokens.doc import Doc
+
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+"""
+Base Class for implementing the different input transformations a generation should be robust against.
+"""
+
+
+def uncapitalize(string: str):
+    """De-capitalize first character of string
+
+    E.g. 'How is Michael doing?' -> 'how is Michael doing?'
+    """
+    if len(string):
+        return string[0].lower() + string[1:]
+    return ''
+
+
+def front_auxiliary(auxiliary: Token) -> str:
+    """Take auxiliary (type: spacy Token) and return capitalized, expanded
+    (i.e. un-contracted) auxiliary (type: str). Differentiates certain English
+    identical English contractions (e.g. "'d", "'s") using morphology data
+    stored in auxiliary `Token` object.
+
+    E.g.:
+    - <Token 'has'> -> 'Has'
+    - <Token "'ve"> -> 'Have'
+    """
+    if auxiliary.text == "'d":
+        if 'Part' in auxiliary.head.morph.get('VerbForm'):
+            return 'Had'
+        else:
+            return 'Would'
+    elif auxiliary.text == "'s":
+        if 'Past' in auxiliary.head.morph.get('Tense'):
+            return 'Has'
+        else:
+            return 'Is'
+    elif auxiliary.text == "'ve":
+        return 'Have'
+    elif auxiliary.text == "'ll":
+        return 'Will'
+    else:
+        return auxiliary.text.capitalize()
+
+
+def front_be_verb(be_verb: Token) -> str:
+    """Take be verb (type: spacy Token), return capitalized, expanded (i.e.
+    un-contracted) form.
+
+    E.g.:
+    - <Token 'is'> -> 'Is'
+    - <Token "'re"> -> 'Are'
+    """
+    if be_verb.text == "'s":
+        return 'Is'
+    elif be_verb.text == "'re":
+        return 'Are'
+    elif be_verb.text == "'m":
+        return 'Am'
+    else:
+        return be_verb.text.capitalize()
+
+
+class YesNoQuestionPerturbation(SentenceOperation):
+    tasks = [
+        TaskType.TEXT_CLASSIFICATION,
+        TaskType.TEXT_TO_TEXT_GENERATION,
+        TaskType.QUESTION_ANSWERING,
+        TaskType.QUESTION_GENERATION
+    ]
+    languages = ["en"]
+
+    def __init__(self, seed=0, max_outputs=1):
+        super().__init__(seed, max_outputs=max_outputs)
+        self.detokenizer = TreebankWordDetokenizer()
+        self.nlp = spacy_nlp if spacy_nlp else spacy.load('en_core_web_sm')
+
+    def statement_to_question(self, sentence: Span) -> Union[str, None]:
+        """Given a statement (type: spacy Span), convert to corresponding
+        yes-or-no question.
+        """
+
+        # FIXME: This is a debugging breadcrumb
+        if sentence.text == "Dave isn't going to Seattle.":
+            raise Exception([(token.text_with_ws, token.dep_, token.pos_,
+                              token.morph) for token in sentence])
+
```

Comment:
Hi @cascadianblue: This exception is causing the build to fail. Please remove it if it's used only  for debugging purpose.